### ISSUE TITLE
possible "qw" bug

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1108]
Migrated from [rt.perl.org#2063](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=2063) (status was 'resolved')

Searchable as RT2063$

**Comment by p5pRT** [comment]
### From dan_liddell@mentorg.com

Possible bug&#8203;:

$a = \("one"\, "two"\, "three"\);
\# $a = three\, as we would expect; however\,

$a = qw\(one two three\);
\# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?

Thanks\!
\-\- 



\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
Mathematics is that subject in which we never
know what we're talking about\, nor whether
what we are saying is true\.

&nbsp;                     \- Bertrand Russell
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*

**Comment by p5pRT** [comment]
### From dan_liddell@mentorg.com

\#\! /usr/bin/perl \-w

$a = \("one"\, "two"\, "three"\);
print "$a\\n";
\# $a = three\, as we would expect; however\,

$a = qw\(one two three\);
print "$a\\n";
\# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?


\-\- 



\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
Mathematics is that subject in which we never
know what we're talking about\, nor whether
what we are saying is true\.

&nbsp;                     \- Bertrand Russell
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

>$a = qw\(one two three\);
>print "$a\\n";
>\# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?

qw\(\) is just shorthand for split ' '\, q\(\);

\-\- 
MIDN 4/C PINYAN\, NROTCURPI\, US Naval Reserve             japhy@&#8203;pobox\.com
http&#8203;://www\.pobox\.com/~japhy/                  http&#8203;://pinyaj\.stu\.rpi\.edu/
PerlMonth \- An Online Perl Magazine            http&#8203;://www\.perlmonth\.com/
The Perl Archive \- Articles\, Forums\, etc\.    http&#8203;://www\.perlarchive\.com/

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

>\#\! /usr/bin/perl \-w

>$a = \("one"\, "two"\, "three"\);
>print "$a\\n";
>\# $a = three\, as we would expect; however\,

>$a = qw\(one two three\);
>print "$a\\n";
>\# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?

% perl5\.005\_63 \-le '$a = qw\(one two three\);  print $a'
three

\-\-tom

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Feb 1\, Tom Christiansen said&#8203;:

>% perl5\.005\_63 \-le '$a = qw\(one two three\);  print $a'
>three

That change had better be documented\.

From 5\.005\_02's perlop\.pod&#8203;:

&nbsp;     qw/STRING/
&nbsp;             Returns a list of the words extracted out of STRING\,
&nbsp;             using embedded whitespace as the word delimiters\.
&nbsp;             It is exactly equivalent to

&nbsp;                 split\(' '\, q/STRING/\);

&nbsp;             This equivalency means that if used in scalar
&nbsp;             context\, you'll get split's \(unfortunate\) scalar
&nbsp;             context behavior\, complete with mysterious warnings\.


\-\- 
MIDN 4/C PINYAN\, NROTCURPI\, US Naval Reserve             japhy@&#8203;pobox\.com
http&#8203;://www\.pobox\.com/~japhy/                  http&#8203;://pinyaj\.stu\.rpi\.edu/
PerlMonth \- An Online Perl Magazine            http&#8203;://www\.perlmonth\.com/
The Perl Archive \- Articles\, Forums\, etc\.    http&#8203;://www\.perlarchive\.com/

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

>>% perl5\.005\_63 \-le '$a = qw\(one two three\);  print $a'
>>three

>That change had better be documented\.

>>From 5\.005\_02's perlop\.pod&#8203;:

>     qw/STRING/
>             Returns a list of the words extracted out of STRING\,
>             using embedded whitespace as the word delimiters\.
>             It is exactly equivalent to

>                 split\(' '\, q/STRING/\);

>             This equivalency means that if used in scalar
>             context\, you'll get split's \(unfortunate\) scalar
>             context behavior\, complete with mysterious warnings\.

Patches to perl\{op\,delta\}\.pod are\, of course\, everwelcome\.

\-\-tom

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Feb 1\, Tom Christiansen said&#8203;:

>>     qw/STRING/
>>             It is exactly equivalent to
>
>>                 split\(' '\, q/STRING/\);
>
>Patches to perl\{op\,delta\}\.pod are\, of course\, everwelcome\.

Patches to perl\{op\,delta\}\.pod will be\, of course\, written\, when I'm told
the new exact equivelent of qw\(\)\, and/or the behavior of split\(\) in scalar
context\.

\-\- 
MIDN 4/C PINYAN\, NROTCURPI\, US Naval Reserve             japhy@&#8203;pobox\.com
http&#8203;://www\.pobox\.com/~japhy/                  http&#8203;://pinyaj\.stu\.rpi\.edu/
PerlMonth \- An Online Perl Magazine            http&#8203;://www\.perlmonth\.com/
The Perl Archive \- Articles\, Forums\, etc\.    http&#8203;://www\.perlarchive\.com/

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

>On Feb 1\, Tom Christiansen said&#8203;:

>>>     qw/STRING/
>>>             It is exactly equivalent to
>>
>>>                 split\(' '\, q/STRING/\);
>>
>>Patches to perl\{op\,delta\}\.pod are\, of course\, everwelcome\.

>Patches to perl\{op\,delta\}\.pod will be\, of course\, written\, when I'm told
>the new exact equivelent of qw\(\)\, and/or the behavior of split\(\) in scalar
&nbsp;                    ^^^^^ Are we talking equal\-velars or equal\-values???
>context\.

From perldelta\.pod&#8203;:

&nbsp;    =head2 Improved C\<qw//> operator

&nbsp;    The C\<qw//> operator is now evaluated at compile time into a
&nbsp;    true list instead of being replaced with a run time call to
&nbsp;    C\<split\(\)>\.  This removes the confusing misbehaviour of C\<qw//>
&nbsp;    in scalar context\, which had inherited that behaviour from
&nbsp;    split\(\)\.

&nbsp;    Thus&#8203;:

&nbsp;        $foo = \($bar\) = qw\(a b c\); print "$foo|$bar\\n";

&nbsp;    now correctly prints "3|a"\, instead of "2|a"\.

And from perlop\.pod&#8203;:

&nbsp;    =item qw/STRING/

&nbsp;    Evaluates to a list of the words extracted out of STRING\, using
&nbsp;    embedded whitespace as the word delimiters\.  It can be understood
&nbsp;    as being roughly equivalent to&#8203;:

&nbsp;        split\(' '\, q/STRING/\);

&nbsp;    the difference being that it generates a real list at compile
&nbsp;    time\.  So this expression&#8203;:

&nbsp;        qw\(foo bar baz\)

&nbsp;    is semantically equivalent to the list&#8203;:

&nbsp;        'foo'\, 'bar'\, 'baz'

In summary\, I believe that one would find that Sarathy has already
done the necessary docwork\, were one but to read it\. :\-\(

\-\-tom

**Comment by p5pRT** [comment]
### From @TimToady

Dan Liddell writes&#8203;:
: \# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?

Nothing ever returns a list in scalar context\, so there's no reason to
expect qw\(\) to do any particular thing\.  I'd say changing it to return
the last value is probably a bit of a flub\.  A good argument could be
made that it should return join\(' '\, qw\(one two three\)\)\.  At least then
it would have the useful function of turning arbitrary whitespace into
consistent whitespace\.

Or maybe it should be join\($"\, qw\(one two three\)\)\.

Larry

**Comment by p5pRT** [comment]
### From @TimToady

Larry Wall writes&#8203;:
: Dan Liddell writes&#8203;:
: : \# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?
: 
: Nothing ever returns a list in scalar context\, so there's no reason to
: expect qw\(\) to do any particular thing\.  I'd say changing it to return
: the last value is probably a bit of a flub\.  A good argument could be
: made that it should return join\(' '\, qw\(one two three\)\)\.  At least then
: it would have the useful function of turning arbitrary whitespace into
: consistent whitespace\.
: 
: Or maybe it should be join\($"\, qw\(one two three\)\)\.

Before you guys leap to the \(probably justifiable\) conclusion that I'm
totally off my rocker\, I should point out that there's precedent in qx//\.

Actually\, I'm just waiting for Unicode to take over so we can switch
to using those cute French angle quotes for qw\(\)\, which I've always thought
was rather ugly\.  Then we'll be debating the meaning of&#8203;:

&nbsp;    $a = «one two three»;

Larry

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Larry Wall \<larry@&#8203;wall\.org> writes&#8203;:

> Or maybe it should be join\($"\, qw\(one two three\)\)\.

For what it's worth\, I like this idea\, and can even see some use for it\.

\-\- 
Russ Allbery \(rra@&#8203;stanford\.edu\)         \<URL&#8203;:http&#8203;://www\.eyrie\.org/~eagle/>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


From my own perspective there is a certain appeal to having qw\(a b c\) be
equivalent to \("a"\, "b"\, "c"\) up to isomorphism\, but then again\, what do
I know? The statement at the end of this message applies equally well to
perl\.


Larry Wall wrote&#8203;:
> 
> Larry Wall writes&#8203;:
> : Dan Liddell writes&#8203;:
> : : \# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?
> :
> : Nothing ever returns a list in scalar context\, so there's no reason to
> : expect qw\(\) to do any particular thing\.  I'd say changing it to return
> : the last value is probably a bit of a flub\.  A good argument could be
> : made that it should return join\(' '\, qw\(one two three\)\)\.  At least then
> : it would have the useful function of turning arbitrary whitespace into
> : consistent whitespace\.
> :
> : Or maybe it should be join\($"\, qw\(one two three\)\)\.
> 
> Before you guys leap to the \(probably justifiable\) conclusion that I'm
> totally off my rocker\, I should point out that there's precedent in qx//\.
> 
> Actually\, I'm just waiting for Unicode to take over so we can switch
> to using those cute French angle quotes for qw\(\)\, which I've always thought
> was rather ugly\.  Then we'll be debating the meaning of&#8203;:
> 
>     $a = «one two three»;
> 
> Larry

\-\- 



\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
Mathematics is that subject in which we never
know what we're talking about\, nor whether
what we are saying is true\.

&nbsp;                     \- Bertrand Russell
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*

**Comment by p5pRT** [comment]
### From @TimToady

Dan Liddell writes&#8203;:
: From my own perspective there is a certain appeal to having qw\(a b c\) be
: equivalent to \("a"\, "b"\, "c"\) up to isomorphism\, but then again\, what do I
: know\.  The statement at the end of this message applies equally well to perl\.

Oh\, I dunno\.  Perl is as much about religion as it is about mathematics\,
and Bertrand Russell didn't seem to have any qualms about making any number
of absolute pronouncements about religion\.  So go for it\.  :\-\)

Larry

: \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
: Mathematics is that subject in which we never
: know what we're talking about\, nor whether
: what we are saying is true\.
: 
:                      \- Bertrand Russell
: \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Larry Wall wrote&#8203;:
> 
> Larry Wall writes&#8203;:
> : Dan Liddell writes&#8203;:
> : : \# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?
> :
> : Nothing ever returns a list in scalar context\, so there's no reason to
> : expect qw\(\) to do any particular thing\.  I'd say changing it to return
> : the last value is probably a bit of a flub\.  A good argument could be
> : made that it should return join\(' '\, qw\(one two three\)\)\.  At least then
> : it would have the useful function of turning arbitrary whitespace into
> : consistent whitespace\.
> :
> : Or maybe it should be join\($"\, qw\(one two three\)\)\.
> 
> Before you guys leap to the \(probably justifiable\) conclusion that I'm
> totally off my rocker\, I should point out that there's precedent in qx//\.

How about

Warning&#8203;: I don't think that means what you think it means\.

Also\, maybe the
$x = \(\) = \<insert list returning expr with odd \!wantarray behavior here>
trick should be smeared around the documentation?

**Comment by p5pRT** [comment]
### From @iabyn

Just reviewing old bugs from the Perl bug database&#8203;:

> Possible bug&#8203;:
> 
> $a = \("one"\, "two"\, "three"\);
> \# $a = three\, as we would expect; however\,
> 
> $a = qw\(one two three\);
> \# $a = 3\, hmm\.\.\. unexpected\, does qw REALLY return a list?

This has been fixed in more recent releases of Perl\, such 5\.6\.1

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'stalled' to 'resolved'
